### PR TITLE
MergeTree: Unify removed client id and overlap clients

### LINKED ISSUE #9138: "Merge tree insert failed" in telemetry
We're seeing some instances of "Merge tree insert failed" errors in telemetry at low volume (ICM 288583185).  Opening this bug to track investigation of these issues.

### PR DESCRIPTION
**Pull request by anthony-murphy**
This change unifies the removedClientId and removedClientOverlap into a single array, removedClientIds.

This simplifies our handling in a number of places, as we no longer need to check and manage both. In fact, there were a couple bugs where both were not handled appropriately. 

In addition, I've improved how we handle IRemovalInfo such that it is now all or nothing, which removes the need for a number of unsafe non-null assertions, and make the code simpler by removing the need to repeated non-null checks.

this may fix #9138, but it hard to know as we don't have much telemetry around the issue.

### CODE REVIEW

File: packages/dds/merge-tree/src/mergeTree.ts [hunk 1]
```diff
@@ -523,7 +529,7 @@ export abstract class BaseSegment extends MergeNode implements ISegment {
         b.clientId = this.clientId;
         // TODO: deep clone properties
         b.properties = clone(this.properties);
-        b.removedClientId = this.removedClientId;
+        b.removedClientIds = this.removedClientIds?.slice()
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/src/mergeTree.ts:532]: there was a bug here where we didn't clone overlap clients

File: packages/dds/merge-tree/src/snapshotLoader.ts [hunk 2]
```diff
@@ -102,8 +102,16 @@ export class SnapshotLoader {
             if (spec.removedSeq !== undefined) {
                 seg.removedSeq = spec.removedSeq;
             }
+            // this format had a bug where it didn't store all the overlap clients
+            // this is for back compat, so we change the singular id to an array
+            // this will only cause problems if there is an overlapping delete
+            //spanning the snapshot, which should be rare
             if (spec.removedClient !== undefined) {
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/src/snapshotLoader.ts:109]: this was another bug where we missed the overlap clients in the snapshot

**Comment by heliocliu** [comment]
Does this change need to go into the 0.58 release? If so, please merge it before Wednesday and let me know so we can copy it over. Otherwise please rebase this against and merge into the `next` branch instead because it is a breaking change.

**Comment by anthony-murphy** [comment]
> Does this change need to go into the 0.58 release? If so, please merge it before Wednesday and let me know so we can copy it over. Otherwise please rebase this against and merge into the `next` branch instead because it is a breaking change.

this change already targets the next branch

**Comment by heliocliu** [comment]
> > Does this change need to go into the 0.58 release? If so, please merge it before Wednesday and let me know so we can copy it over. Otherwise please rebase this against and merge into the `next` branch instead because it is a breaking change.
> 
> this change already targets the next branch

My bad, I missed that 😳 Thanks for being on top of things!

File: packages/dds/merge-tree/src/mergeTree.ts [hunk 3]
```diff
@@ -104,15 +104,22 @@ export interface IHierBlock extends IMergeBlock {
...
 
 export interface IRemovalInfo {
-    removedSeq?: number;
-    removedClientId?: number;
-    removedClientOverlap?: number[];
+    removedSeq: number;
+    removedClientIds: number[];
+}
+export function toRemovalInfo(maybe: Partial<IRemovalInfo> | undefined): IRemovalInfo | undefined{
+
```
Comment (pleath) [inline packages/dds/merge-tree/src/mergeTree.ts:111]: If maybe is a Partial instance, is it possible for just one of removedClientIds and removedSeq to be undefined? Is it possible for maybe to be passed as "IRemovalInfo | undefined"?

Comment (anthony-murphy) [reply packages/dds/merge-tree/src/mergeTree.ts:111]: this function will ensure either none or both at set. generally, what we pass in is ISegment, which extends `Partial<IRemovalInfo>` as it only sets those on removal. in the future with more refactorings we may be able to tighten the typing but i think right now this is the best we can do.

**Comment by anthony-murphy** [comment]
related to #6957 #9400

File: packages/dds/merge-tree/src/mergeTree.ts (hunk 3 above)
Comment (pleath) [reply packages/dds/merge-tree/src/mergeTree.ts:111]: I see. Thanks.

**Review by pleath — APPROVED** [review]

File: packages/dds/merge-tree/src/snapshotLoader.ts (hunk 2 above)
Comment (DLehenbauer) [reply packages/dds/merge-tree/src/snapshotLoader.ts:109]: My mistake.  Thanks for fixing!

**Review by DLehenbauer — APPROVED** [review]

**Pull request merged**